From 4e6572e679cb23e3b337acedba36ef114f2f4dcb Mon Sep 17 00:00:00 2001
From: Jaeyoon Jung <jaeyoon.jung@lge.com>
Date: Fri, 23 Feb 2018 08:56:31 +0900
Subject: [PATCH] Fix baseUrl returning an empty url

Do not append the directory separator if the current path is the root
directory. Otherwise it is treated as a remote path incorrectly and an
empty url is returned as a result.

Change-Id: Icc62114702265e9e1b247624370a78999e564d5f
Task-number: QTBUG-83121
Reviewed-by: Fabian Kosmale <fabian.kosmale@qt.io>
(cherry picked from commit bba1b19a18a04060942e5580e36daccf9c830b61)
Reviewed-by: Qt Cherry-pick Bot <cherrypick_bot@qt-project.org>
Upstream-Status: Backported [https://codereview.qt-project.org/c/qt/qtdeclarative/+/303662]
---
 src/qml/qml/qqmlengine.cpp                   |  4 +++-
 tests/auto/qml/qqmlengine/tst_qqmlengine.cpp | 13 +++++++++++++
 2 files changed, 16 insertions(+), 1 deletion(-)

diff --git a/src/qml/qml/qqmlengine.cpp b/src/qml/qml/qqmlengine.cpp
index a32a13b5a4..5cb83574b7 100644
--- a/src/qml/qml/qqmlengine.cpp
+++ b/src/qml/qml/qqmlengine.cpp
@@ -1341,7 +1341,9 @@ QUrl QQmlEngine::baseUrl() const
 {
     Q_D(const QQmlEngine);
     if (d->baseUrl.isEmpty()) {
-        return QUrl::fromLocalFile(QDir::currentPath() + QDir::separator());
+        const QString currentPath = QDir::currentPath();
+        const QString rootPath = QDir::rootPath();
+        return QUrl::fromLocalFile((currentPath == rootPath) ? rootPath : (currentPath + QDir::separator()));
     } else {
         return d->baseUrl;
     }
diff --git a/tests/auto/qml/qqmlengine/tst_qqmlengine.cpp b/tests/auto/qml/qqmlengine/tst_qqmlengine.cpp
index ce7f509bc0..329c023d84 100644
--- a/tests/auto/qml/qqmlengine/tst_qqmlengine.cpp
+++ b/tests/auto/qml/qqmlengine/tst_qqmlengine.cpp
@@ -212,6 +212,19 @@ void tst_qqmlengine::baseUrl()
     engine.setBaseUrl(cwd);
     QCOMPARE(engine.baseUrl(), cwd);
     QCOMPARE(engine.rootContext()->resolvedUrl(QUrl("main.qml")), cwd.resolved(QUrl("main.qml")));
+
+
+    const QString testPath = QDir::currentPath() + QLatin1String("/");
+    const QString rootPath = QDir::rootPath();
+    engine.setBaseUrl(QUrl());
+
+    // Check that baseUrl returns a url to a localFile
+    QCOMPARE(engine.baseUrl().toLocalFile(), testPath);
+
+    QDir::setCurrent(QDir::rootPath());
+
+    // Make sure this also works when in the rootPath
+    QCOMPARE(engine.baseUrl().toLocalFile(), rootPath);
 }
 
 void tst_qqmlengine::contextForObject()
